### PR TITLE
MDEV-6284 - Partially merge downstream Debian/Ubuntu packaging into upstream MariaDB

### PR DESCRIPTION
**Pull request by ottok**
The packaging of MariaDB in [official Debian repositories](https://tracker.debian.org/pkg/mariadb-10.0) is almost completely re-written to match modern Debian policy, quality and other requirements. The issue [MDEV-6284](https://jira.mariadb.org/browse/MDEV-6284) has been open since 2014 as the intent has for long been to merge these changes back up into the MariaDB server sources, so that the packages at mariadb.org would be equally good as now in Debian official repositories.

This PR is based on PR #202 that was originally made already in June 2016 but never merged, so I started to refresh and rebase it in early October, and when I found out the new changes in 10.2 I rewrote much of it. This is why some of the commits have old dates from July. Now this PR is rebased on the latest 10.2 branch as of today.

git-history
## Contents

The commit log titles show the major contents of this change. Most importantly the Debian packaging generation is updated to modern dh_\* based rules (but not too modern, so we are still compatible with all current Debian/Ubuntu releases that are still maintained).

This change does _not_ include unix socket authentication nor some default-mysql-\* or default-libmysqlclient -related changes that are in Debian but which are not compatible or relevant to have in our own MariaDB.org repositories.

This PR does not either solve all issues linked to MDEV-6284, but I will continue to solve those in a later PR.

The focus of this changeset is to update and align the Debian packaging, and _not_ to make too invasive changes.
## Quality assurance
- basic Travis-Ci tests pass at https://travis-ci.org/ottok/mariadb/builds/171659578
- builds for architectures amd64, i386 and ppc64 pass at https://launchpad.net/~mysql-ubuntu/+archive/ubuntu/mariadb-10.2/+builds?build_text=&build_state=all (before this PR 10.2 only passes on amd64 on Launchpad)
- buildbot at http://buildbot.askmonty.org/buildbot/grid?branch=ok-debpkg-10.2&category=main&category=experimental is not all green, but it cannot be as the master 10.2 is not green and some of the tests fail due to buildbot timeouts (not faults in this changeset) and some tests are impossible to solve for now (e.g. MySQL 5.7 upgrade as explained in [MDEV-11170](https://jira.mariadb.org/browse/MDEV-11170)).
- Use your favourite diff tool to compare the filelists of 10.2 before and after this changeset. It shows there are no unintentional regressions in files being missed dropped from the packages when restructuring the .install files et al:
  - http://labs.seravo.fi/~otto/debian/mariadb-xenial-amd64/filelist-10.2.2.log
  - http://labs.seravo.fi/~otto/debian/mariadb-xenial-amd64/filelist-c912d05.log

I don't expect anybody to review this PR with 90+ files changed in detail, but I'll file it anyway and let is sit for a few days in case somebody wants to chip in, and then merge it myself to that it willl be included in the release of 10.2.3.

### CODE REVIEW

File: debian/mariadb-plugin-mroonga.prerm [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#!/bin/sh
+
+set -e
+
+# Install Mroonga
+# No user or password parameter is required with new MariaDB that
+# has unix socket authentication support by default.
```
Comment (grooverdan) [inline debian/mariadb-plugin-mroonga.prerm]: ^ this unfortunately isn't the case.

File: debian/control [hunk 2]
```diff
@@ -161,6 +159,7 @@ Conflicts: mariadb-client-10.0,
            mysql-client-core-5.1,
            mysql-client-core-5.5,
            mysql-client-core-5.6,
+           mysql-client-core-5.7,
```
Comment (grooverdan) [inline debian/control]: should mysql-8.0 packages from Oracle be accounted for here?

**Comment by grooverdan** [comment]
Nice work. Well done.

File: debian/control (hunk 2 above)
Comment (ottok) [inline debian/control]: Not yet, we can update that later.

File: debian/mariadb-plugin-mroonga.prerm (hunk 1 above)
Comment (ottok) [inline debian/mariadb-plugin-mroonga.prerm]: Right, comment is not updated after move from Debian.org, but contents has `--defaults-file=/etc/mysql/debian.cnf`

**Comment by ottok** [comment]
I plan to merge this PR in Github today, so the whole chain of commits is visible via one separate merge commit.

**Pull request merged**